php_phongo.c

```diff
@@ -590,48 +590,79 @@ int phongo_execute_query(mongoc_client_t *client, const char *namespace, zval *z
+	opts = bson_new();
+	if (server_id > 0) {
+		bson_append_int32(opts, "serverId", -1, server_id);
+		selected_server_id = server_id;
+	} else {
+		mongoc_server_description_t  *selected_server = NULL;
 
-	if (server_id > 0 && !mongoc_cursor_set_hint(cursor, server_id)) {
-		phongo_throw_exception(PHONGO_ERROR_MONGOC_FAILED TSRMLS_CC, "%s", "Could not set cursor server_id");
-		return false;
+		selected_server = mongoc_client_select_server(client, false, (zreadPreference ? phongo_read_preference_from_zval(zreadPreference TSRMLS_CC) : mongoc_client_get_read_prefs(client)), &error);
+		if (selected_server) {
+			selected_server_id = mongoc_server_description_id(selected_server);
+			bson_append_int32(opts, "serverId", -1, selected_server_id);
+			mongoc_server_description_destroy(selected_server);
+		} else {
+			/* Check for connection related exceptions */
+			if (!EG(exception)) {
+				phongo_throw_exception_from_bson_error_t(&error TSRMLS_CC);
+			}
+
+			bson_free(opts);
+			return false;
+		}
 	}
 
-	if (!phongo_advance_cursor_and_check_for_error(cursor TSRMLS_CC)) {
+	if (!mongoc_client_command_with_opts(client, db, command->bson, phongo_read_preference_from_zval(zreadPreference TSRMLS_CC), opts, &reply, &error)) {
+		phongo_throw_exception_from_bson_error_t(&error TSRMLS_CC);
+		bson_free(opts);
 		return false;
 	}
 
+	bson_free(opts);
+
 	if (!return_value_used) {
-		mongoc_cursor_destroy(cursor);
+		bson_destroy(&reply);
 		return true;
 	}
 
-	if (bson_iter_init_find(&iter, mongoc_cursor_current(cursor), "cursor") && BSON_ITER_HOLDS_DOCUMENT(&iter)) {
-		mongoc_cursor_t *cmd_cursor;
-
-		/* According to mongoc_cursor_new_from_command_reply(), the reply bson_t
-		 * is ultimately destroyed on both success and failure. Use bson_copy()
-		 * to create a writable copy of the const bson_t we fetched above. */
-		cmd_cursor = mongoc_cursor_new_from_command_reply(client, bson_copy(mongoc_cursor_current(cursor)), mongoc_cursor_get_hint(cursor));
-		mongoc_cursor_destroy(cursor);
+	/* According to mongoc_cursor_new_from_command_reply(), the reply bson_t
+	 * is ultimately destroyed on both success and failure. */
+	if (bson_iter_init_find(&iter, &reply, "cursor") && BSON_ITER_HOLDS_DOCUMENT(&iter)) {
+		cmd_cursor = mongoc_cursor_new_from_command_reply(client, &reply, selected_server_id);
+	} else {
+		bson_t *wrapped_reply = create_wrapped_command_envelope(&reply);
 
-		if (!phongo_advance_cursor_and_check_for_error(cmd_cursor TSRMLS_CC)) {
-			return false;
-		}
+		cmd_cursor = mongoc_cursor_new_from_command_reply(client, wrapped_reply, selected_server_id);
+		bson_destroy(&reply);
+	}
 
-		phongo_cursor_init_for_command(return_value, client, cmd_cursor, db, zcommand, zreadPreference TSRMLS_CC);
-		return true;
+	if (!phongo_advance_cursor_and_check_for_error(cmd_cursor TSRMLS_CC)) {
+		return false;
```

Comment:
Should we call `mongoc_cursor_destroy(cmd_cursor);` before returning here? There could be a leak, and I don't think we have any tests that cover an error during the first advance.

On a related note, `phongo_execute_query()` may also have two possible leaks in the following code blocks:

```c
if (server_id > 0 && !mongoc_cursor_set_hint(cursor, server_id)) {
	phongo_throw_exception(PHONGO_ERROR_MONGOC_FAILED TSRMLS_CC, "%s", "Could not set cursor server_id");
	/* THIS MAY LEAK */
	return false;
}

/* maxAwaitTimeMS must be set before the cursor is sent */
if (query->max_await_time_ms) {
	mongoc_cursor_set_max_await_time_ms(cursor, query->max_await_time_ms);
}

if (!phongo_advance_cursor_and_check_for_error(cursor TSRMLS_CC)) {
	/* THIS MAY LEAK */
	return false;
}
```

Reply:
The ones for ``phongo_execute_query`` should probably go into a new PR, however, we'll probably have to think hard on how to make this fail within a test case.

I've updated it here.

Reply:
I'll open a ticket for `phongo_execute_query`: [PHPC-1043](https://jira.mongodb.org/browse/PHPC-1043).